File: ltech/app/src/main/java/merryweather/com/ltech/di/ComponentManager.java
package merryweather.com.ltech.di;

import android.app.Activity;
import android.app.Application;
import android.support.annotation.NonNull;

import java.util.HashMap;
import java.util.Map;

import merryweather.com.ltech.App;
import merryweather.com.ltech.di.component.ActivityComponent;
import merryweather.com.ltech.di.component.AppComponent;
import merryweather.com.ltech.di.component.DaggerActivityComponent;
import merryweather.com.ltech.di.component.DaggerAppComponent;
import merryweather.com.ltech.di.modules.ActivityModule;
import merryweather.com.ltech.di.modules.AppModule;


/**
 * Created by S on 19.04.2018.
 */

public final class ComponentManager {

    private AppComponent appComponent;

    private final Map<String, ActivityComponent> activityComponentMap = new HashMap<>();


    private ComponentManager() {
    }

    @NonNull
    public static ComponentManager getInstance() {
        return Holder.INSTANCE;
    }

    public final void init(final Application application) {
        appComponent = DaggerAppComponent.builder()
                .appModule(new AppModule(application))
                .build();

        appComponent.inject((App) application);
    }

    /**
     * For Services, Broadcasts, etc.
     */
    public final AppComponent getAppComponent() {
        return appComponent;
    }

    /**
     * For Activities
     */
    public final ActivityComponent getActivityComponent(final Activity activity) {
        final String key = getKey(activity);
        ActivityComponent activityComponent = activityComponentMap.get(key);
        if (activityComponent == null) {
            activityComponent = createNewActivityComponent(activity);
            activityComponentMap.put(key, activityComponent);
        }
        return activityComponent;
    }

    public final void removeActivityComponent(final Activity activity) {
        activityComponentMap.remove(getKey(activity));
    }


    private ActivityComponent createNewActivityComponent(final Activity activity) {
        return DaggerActivityComponent.builder()
                .appComponent(appComponent)
                .activityModule(new ActivityModule(activity))
                .build();
    }

    @NonNull
    private String getKey(final Activity activity) {
        return activity.getLocalClassName();
    }

    private static final class Holder {
        private static final ComponentManager INSTANCE = new ComponentManager();
    }

}

File: Nista/app/src/main/java/com/example/nista/di/ComponentManager.kt
package com.example.nista

import android.app.Activity
import android.app.Application
import android.app.Service
import androidx.fragment.app.Fragment
import com.example.nista.di.component.*
import com.example.nista.di.modules.FragmentModule

import java.util.HashMap

import com.leroymerlin.cds.di.modules.ActivityModule
import com.leroymerlin.cds.di.modules.AppModule


class ComponentManager private constructor() {

    /**
     * For Services, Broadcasts, etc.
     */
    lateinit var appComponent: AppComponent
        private set

    private val activityComponentMap = HashMap<String, ActivityComponent>()

    private val fragmentComponentMap = HashMap<String, FragmentComponent>()

    fun init(application: Application) {
        appComponent = DaggerAppComponent.builder()
                .appModule(AppModule(application))
                .build()

        appComponent.inject(application as App)
    }

    /**
     * For Activities
     */
    fun getActivityComponent(activity: Activity): ActivityComponent {
        val key = getKey(activity)
        var activityComponent: ActivityComponent? = activityComponentMap[key]
//        if (activityComponent == null) {
        activityComponent = createNewActivityComponent(activity)
        activityComponentMap[key] = activityComponent
//        }
        return activityComponent
    }

    fun removeActivityComponent(activity: Activity) {
        activityComponentMap.remove(getKey(activity))
    }


    private fun createNewActivityComponent(activity: Activity): ActivityComponent {
        return DaggerActivityComponent.builder()
                .appComponent(appComponent)
                .activityModule(ActivityModule(activity))
                .build()
    }

    /**
     * For Fragments
     */
    fun getFragmentComponent(fragment: Fragment): FragmentComponent {
        val activity = fragment.requireActivity()
        val args = fragment.arguments
        val key_suffix = if (args != null) "" + args.hashCode() else ""
        val key = getKey(activity) + key_suffix
        val activityComponent = getActivityComponent(activity)

//        if (fragmentComponent == null) {
        val fragmentComponent = DaggerFragmentComponent.builder()
                .activityComponent(activityComponent)
                .fragmentModule(FragmentModule(fragment))
                .build()
        fragmentComponentMap.put(key, fragmentComponent)
//        }
        return fragmentComponent

    }

    fun removeFragmentComponent(fragment: Fragment) {
        val activity = fragment.requireActivity()
        val args = fragment.arguments
        val key_suffix = if (args != null) "" + args.hashCode() else ""
        val key = getKey(activity) + key_suffix

        fragmentComponentMap.remove(key)
    }

    private fun getKey(activity: Activity): String {
        return activity.localClassName
    }

    private object Holder {
        val INSTANCE = ComponentManager()
    }

    companion object {

        val instance: ComponentManager by lazy { Holder.INSTANCE }
    }

}